zulip/zulip/__init__.py

```diff
@@ -667,10 +674,15 @@ def end_error_retry(succeeded: bool) -> None:
                 if error_retry(""):
                     continue
                 end_error_retry(False)
-                return {
-                    "msg": f"Connection error:\n{traceback.format_exc()}",
-                    "result": "connection-error",
-                }
+
+                msg = f"Connection error:\n{traceback.format_exc()}"
+                if self.legacy_exception_handling:
+                    return {
+                        "msg": msg,
+                        "result": "connection-error",
+                    }
+                else:
+                    raise UnrecoverableNetworkError(msg)
             except Exception:
                 # We'll split this out into more cases as we encounter new bugs.
                 return {
```

Comment:
This `return` also needs to be addressed.

Reply:
Done :)